src/main/scala/monix/nio/tcp/AsyncServerSocketChannel.scala

```diff
@@ -0,0 +1,173 @@
+ * @define localDesc the local address to bind the socket, or null to bind to an automatically assigned socket address
+ * @define backlogDesc the maximum number of pending connections. If the backlog parameter has the value 0,
+ *         or a negative value, then an implementation specific default is used.
+ *
+ * @define localAddressDesc Asks the socket address that this channel's socket is bound to
+ */
+abstract class AsyncServerSocketChannel extends AutoCloseable {
+
+  /**
+   * $acceptDesc
+   *
+   * @param cb $callbackDesc
+   */
+  def accept(cb: Callback[AsyncSocketChannel]): Unit
+
+  /**
+   * $acceptDesc
+   */
+  def accept(): Future[AsyncSocketChannel] = {
+    val p = Promise[AsyncSocketChannel]()
+    accept(Callback.fromPromise(p))
+    p.future
+  }
+
+  /**
+   * $acceptDesc
+   */
+  def acceptL(): Task[AsyncSocketChannel] =
+    Task.unsafeCreate { (context, cb) =>
+      implicit val s = context.scheduler
+      accept(Callback.async(cb))
+    }
+
+  /**
+   * $bindDesc
+   *
+   * @param local $localDesc
+   * @param backlog $backlogDesc
+   */
+  def bind(local: InetSocketAddress, backlog: Int = 0): Unit
+
+  /**
+   * $localAddressDesc
+   */
+  def localAddress(): Option[InetSocketAddress]
+}
+
+object AsyncServerSocketChannel {
+  /**
+   * Opens a server-socket channel for the given [[java.net.InetSocketAddress]]
+   *
+   * @param reuseAddress [[java.net.ServerSocket#setReuseAddress]]
+   * @param receiveBufferSize [[java.net.Socket#setReceiveBufferSize]] [[java.net.ServerSocket#setReceiveBufferSize]]
+   *
+   * @param s is the `Scheduler` used for asynchronous computations
+   *
+   * @return an [[monix.nio.tcp.AsyncServerSocketChannel]] instance for handling connections.
+   */
+  def apply(
+    reuseAddress: Boolean = true,
+    receiveBufferSize: Int = 256 * 1024
+  )(implicit s: Scheduler): AsyncServerSocketChannel = {
+
+    NewIOImplementation(reuseAddress, receiveBufferSize)
+  }
+
+  private final case class NewIOImplementation(
+      reuseAddress: Boolean = true,
+      receiveBufferSize: Int = 256 * 1024
+  )(implicit scheduler: Scheduler) extends AsyncServerSocketChannel {
+
+    private[this] lazy val asynchronousServerSocketChannel: Either[Throwable, AsynchronousServerSocketChannel] = try {
+      val ag = AsynchronousChannelGroup.withThreadPool(ExecutorServiceWrapper(scheduler))
+      val server = AsynchronousChannelProvider.provider().openAsynchronousServerSocketChannel(ag)
+
+      server.setOption[java.lang.Boolean](StandardSocketOptions.SO_REUSEADDR, reuseAddress)
+      server.setOption[Integer](StandardSocketOptions.SO_RCVBUF, receiveBufferSize)
+
+      Right(server)
+    } catch {
+      case NonFatal(exc) =>
+        scheduler.reportFailure(exc)
+        Left(exc)
+    }
+
+    override def accept(cb: Callback[AsyncSocketChannel]): Unit = {
+      val handler = new CompletionHandler[AsynchronousSocketChannel, Null] {
```

Comment:
Creating this `CompletionHandler` every time is a little wasteful, because the necessary `Callback` can be passed in the `attachment` (here you're using `Null`). So you can create a reusable instance:

```scala
object AsyncServerSocketChannel {
  // ...

  private final val acceptHandler: CompletionHandler[AsynchronousSocketChannel, Callback[AsyncSocketChannel]] = ???
}
``` 

This matters because it means less object allocations and if the API allows us to do this, then why not?